File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    int[] dx = {1, 0, -1, 0};
    int[] dy = {0, 1, 0, -1};
    
    public int solution(int[][] maps) {
        int answer = 0;
        
        int[][] visited = new int[maps.length][maps[0].length];
        
        bfs(maps, visited);
        answer = visited[maps.length-1][maps[0].length-1];
        
        if(answer == 0){
            answer = -1;
        }
        
        return answer;
    }
    
    public void bfs(int[][] maps, int[][] visited){
        visited[0][0] = 1;
        Queue<int[]> q = new LinkedList<int[]>();
        q.offer(new int[]{0,0});
        
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            
            for(int i=0; i<4; i++) {
                int nx = x+dx[i];
                int ny = y+dy[i];
                if (0 <= nx & nx < maps[0].length & 0 <= ny & ny < maps.length) {
                    if (visited[ny][nx] == 0 && maps[ny][nx] == 1){
                        q.offer(new int[]{nx,ny});
                        visited[ny][nx] = visited[y][x] + 1;
                    }
                }
                
            }
        }
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
dx = [0,0,1,-1]
dy = [1,-1,0,0]

# def solution(maps):
#     answer = 0
#     n, m = len(maps), len(maps[0])
#     visited = [[0]*m for _ in range(n)]
#     queue = deque()
#     queue.append([0,0])
#     visited[0][0] = 1
    
#     while queue:
#         x,y = queue.popleft()
        
#         for i in range(4):
#             nx, ny = x+dx[i], y+dy[i]
#             if 0<=nx<m and 0<=ny<n and visited[ny][nx] == 0:
#                 if maps[ny][nx] == 1:
#                     queue.append([nx,ny])
#                     visited[ny][nx] = visited[y][x] + 1

                        
#     if visited[n-1][m-1] != 0:
#         return visited[n-1][m-1]
#     else:
#         return -1

def solution(maps):
    n,m = len(maps), len(maps[0])
    visited = [[0]*m for _ in range(n)]
    queue = deque([(0,0)])
    visited[0][0] = 1
    
    while queue:
        x,y = queue.popleft()
        print(x,y)
        
        for i in range(4):
            nx, ny = x+dx[i], y+dy[i]
            
            if 0<= nx < m and 0 <= ny < n and visited[ny][nx] == 0 and maps[ny][nx] == 1:
                queue.append([nx,ny])
                visited[ny][nx] = visited[y][x] + 1
                
    if visited[n-1][m-1] != 0:
        return visited[n-1][m-1]
    else:
        return -1
    
    